docs/madr/decisions/011-mesh-traffic-route.md

```diff
@@ -0,0 +1,964 @@
+set of `backendRefs` which has negative consequences including but not limited to
+metrics handling and a potentially large number of clusters.
+
+#### Merging behavior
+
+With `targetRef` resources, the ultimate configuration of a data plane proxy
+is always a result of all policies targeting that proxy depending on their
+specificity.
+
+> :question: Is the current merge behavior what we need for routing policy? What surprises are there? See the next sections
+
+This proposal addresses one issue with a "naive" application of `targetRef`
+structure to a routing policy.
+
+##### Surprising behavior of top level `default`
+
+Let's look at two basic theoretical route policies where we include `default` at
+the top level of a `to` rule:
+
+```yaml
+metadata:
+ name: owner
+spec:
+ targetRef:
+  kind: Mesh
+ to:
+  - targetRef:
+     kind: MeshService
+     name: backend
+    default:
+     rules:
+      - matches:
+         path:
+          prefix: /v1
+        backendRefs:
+         - weight: 100
+           kind: MeshServiceSubset
+           name: backend
+           tags:
+            version: v1
+      - matches:
+         path:
+          prefix: /v2
+        backendRefs:
+         - weight: 100
+           kind: MeshServiceSubset
+           name: backend
+           tags:
+            version: v2
+---
+metadata:
+ name: consumer
+spec:
+ targetRef:
+  kind: MeshService
+  name: frontend
+ to:
+  - targetRef:
+     kind: MeshService
+     name: backend
+    default:
+     rules:
+      - matches:
+         path:
+          prefix: /v2
+        filters:
+         - requestHeaderModifier:
+            add:
+             - name: env
+               value: dev
+        backendRefs:
+         - weight: 100
+           kind: MeshServiceSubset
+           name: backend
+           tags:
+            version: v2
+```
+
+The rules for merging a field in a `targetRef` policy is that more specific
+routes replace the value from less specific routes. So the final value of `rules`
+is the value from the `consumer` route:
+
+```yaml
+spec:
+ targetRef:
+  kind: MeshService
+  name: frontend
+ to:
+  - targetRef:
+     kind: MeshService
+     name: backend
+    rules:
+     - matches:
+        path:
+         prefix: /v2
+       filters:
+        - requestHeaderModifier:
+           add:
+            - name: env
+              value: dev
+       backendRefs:
+        - weight: 100
+          kind: MeshServiceSubset
+          name: backend
+          tags:
+           version: v2
+```
+
+This is likely surprising, especially because `consumer` didn't specify anything
+at all for `/v1`.
+
+##### Solution
+
+This MADR proposes to distribute `default` down into individual `rules` and merge
+`rules` based on _structural equality_ of the `matches` value.
+So we would instead have:
+
+```yaml
+metadata:
+ name: owner
+spec:
+ targetRef:
+  kind: Mesh
+ to:
+  - targetRef:
+     kind: MeshService
+     name: backend
+    rules:
```

Comment:
What do you think about changing the name `rules` to something else? We use the word `rules` in the context of building a rule-based view for the `from` list (in MeshTrafficPermission). I think this can create confusion since these are different concepts. We can name this list `routes`. Or we can name the list `matches` and items inside rename to `match`. 
```yaml
to:
  - targetRef:
      kind: MeshService
      name: backend
    routes:
      - matches:
          path:
            prefix: /v1
      - matches:
          path:
            prefix: /v2
```
or
```YAML
to:
  - targetRef:
      kind: MeshService
      name: backend
    matches:
      - match:
          path:
            prefix: /v1
      - match:
          path:
            prefix: /v2
```

Reply:
I prefer routes to matches

Reply:
rules comes from gateway API of course but :shrug:

Reply:
So maybe sticking with rules is better. After all sure it's rules in both cases but they are significantly different things